### PR TITLE
Result challenges on Variant/js

### PR DESCRIPTION
**Pull request by thomas-ode**
Hi,

I tried to run the unit tests but it seems some methods were not implemented or have bugs in them.

I implemented the code so that the tests pass.

Regards, Thomas

### CODE REVIEW

File: 00-max.js [hunk 1]
```diff
@@ -1,3 +1,5 @@
 export function max(array) {
     // TODO: implement max function
```
Comment (jeremyVignelles) [inline 00-max.js]: This TODO can now be removed :)

File: 01-report.js [hunk 2]
```diff
@@ -12,14 +12,18 @@ export function getLast12MonthReport(database) {
     // Also, reading the doc might be helpful :D
     const today = new Date();
     const d = new Date();
+    if(today.getDate() === 29 && today.getMonth() === 1) {
+        today.setDate(10);
+        d.setDate(10);
+    }
```
Comment (jeremyVignelles) [inline 01-report.js]: Feels like a hack, but is correct. Values seems to be picked randomly.
I think that instead of using "today" as a reference, I would take "firstDayOfMonth", by using `setDate(1)` always.
The first day of the month is always defined, no matter the month we chose, so we can now iterate on the correct month.

I tend to avoid "edge cases" like "having to handle the 29th day of february" separately, because that's a path of code that rarely gets called. Sometimes, you can't avoid them, and you have several risks:
- Your code wasn't tested enough and you have a bug that occurs once every four years
- Someone after you edits the code, and introduce a bug in said edge case
- Someone after you doesn't understand what this code is for, removes it, tests, and doesn't notice he's introducing back an issue that was fixed specifically by this code.

You can do two things to mitigate that:
- Comment such code to avoid deletion
- Create unit tests (and make sure devs actually runs them before sending to production 😄) to detect regressions.

File: 02-flatten-array.js [hunk 3]
```diff
@@ -7,15 +7,16 @@
...
+    // for (let i = 0; i < arr.length; i++) {
+    //     if (Array.isArray(arr[i])) {
+    //         for (let j = 0; j < arr[i].length; j++) {
+    //             result.push(arr[i][j]);
+    //         }
+    //     } else {
+    //         result.push(arr[i]);
+    //     }
+    // }
+    const result = arr.flat(Infinity);
```
Comment (jeremyVignelles) [inline 02-flatten-array.js]: The customer updated its specifications : The use of arr.flat() is prohibited because they use an old browser that doesn't support it. Could you please rebase your work ?

(But OK, 10 points to griffindor for telling me about this method 😂 )

**Review by jeremyVignelles — COMMENTED** [review]
Thanks for your submission, you are on the right track, but I'm sure you can improve your code a little bit.
I've improved the unit tests to make sure your code is OK. Can you rebase your branch ?
Also, I don't know why the CI is failing if it works on your machine... We'll have to fix that later...

File: 01-report.js [hunk 4]
```diff
@@ -5,21 +5,18 @@
...
 export function getLast12MonthReport(database) {
-    // TODO: This method does not work, I don't know why.
-    // Please have a look, I really don't see anything wrong.
-    // Note: This one is a bit hard, you will need to use the debugger to understand what is going on, or use the console.log.
-    // Please don't get stuck too long on this one, ask for help if you need it.
-    // Also, reading the doc might be helpful :D
     const today = new Date();
     const d = new Date();
-    const year = d.getFullYear();
+    //Hack to set the day to be the 2nd of the month on normal years, and the first on leap yars
```
Comment (jeremyVignelles) [inline 01-report.js:10]:
```suggestion
    //Hack to set the day to be the 2nd of the month to avoid issues with leap years (february 29th doesn't exist one year before)
```

**Review by jeremyVignelles — APPROVED** [review]
Nice work, thanks for your submission :) You know how to find bugs and fix them.
